File: scgraph/transit_node_routing.py
import json
from heapq import heappush, heappop
from typing import Any, Optional
from scgraph.contraction_hierarchies import CHGraph


class TNRGraphIO:
    def save_as_tnrjson(self, filename: str) -> None:
        """
        Function:
        - Save the current TNRGraph as a JSON file

        Required Arguments:
        - `filename`
            - Type: str
            - What: The filename to save the JSON file as
            - Note: The filename must end with .tnrjson
        """
        if not filename.endswith(".tnrjson"):
            raise ValueError("Filename must end with .tnrjson")

        # Convert tuple keys in distance_table to strings for JSON
        dist_table_str = {
            f"({k[0]},{k[1]})": v for k, v in self.distance_table.items()
        }

        # Convert access node keys to strings
        f_access_str = [
            {str(k): v for k, v in d.items()} for d in self.forward_access_nodes
        ]
        b_access_str = [
            {str(k): v for k, v in d.items()}
            for d in self.backward_access_nodes
        ]

        # Shortcuts also need string keys
        shortcuts_str = {
            str(key): via_node_id for key, via_node_id in self.shortcuts.items()
        }

        with open(filename, "w") as f:
            json.dump(
                {
                    "type": "TNRGraph",
                    "nodes_count": self.nodes_count,
                    "transit_nodes": list(self.transit_nodes),
                    "distance_table": dist_table_str,
                    "forward_access_nodes": f_access_str,
                    "backward_access_nodes": b_access_str,
                    "ch_data": {
                        "ranks": self.ranks,
                        "forward_graph": self.forward_graph,
                        "backward_graph": self.backward_graph,
                        "shortcuts": shortcuts_str,
                        "original_graph": self.original_graph,
                        "nodes_count": self.nodes_count,
                    },
                },
                f,
            )

    @classmethod
    def load_from_tnrjson(cls, filename: str) -> "TNRGraph":
        """
        Function:
        - Load a TNRGraph from a JSON file

        Required Arguments:
        - `filename`
            - Type: str
            - What: The filename of the JSON file to load
        """
        if not filename.endswith(".tnrjson"):
            raise ValueError("Filename must end with .tnrjson")

        with open(filename, "r") as f:
            data = json.load(f)

        if data.get("type") != "TNRGraph":
            raise ValueError("JSON file is not a valid TNRGraph.")

        data.pop("type")
        return cls(**data)


class TNRGraphPreprocessing:
    def __compute_access_nodes__(
        self, node_id: int, forward: bool = True
    ) -> dict[int, float]:
        """
        Function:
        - Perform a restricted upward CH search to find the first transit nodes encountered
        """
        access_nodes = {}
        distances = {node_id: 0}
        open_leaves = [(0, node_id)]

        while open_leaves:
            dist, current_id = heappop(open_leaves)
            if dist > distances.get(current_id, float("inf")):
                continue

            # If we hit a transit node, we stop searching this branch
            if current_id in self.transit_nodes:
                if dist < access_nodes.get(current_id, float("inf")):
                    access_nodes[current_id] = dist
                continue

            # Check neighbors in the upward graph
            # If current_id is an added node (rank inf), it uses original_graph logic
            # (handled by inheriting from CHGraph)
            neighbors = self.__get_neighbors__(current_id, forward)

            for neighbor_id, weight in neighbors.items():
                new_dist = dist + weight
                if new_dist < distances.get(neighbor_id, float("inf")):
                    distances[neighbor_id] = new_dist
                    heappush(open_leaves, (new_dist, neighbor_id))
        return access_nodes

    def __preprocess_tnr__(self, num_transit_nodes: int):
        """
        Function:
        - Perform TNR preprocessing on top of an existing CH
        """
        # 1. Select Transit Nodes (Top N by rank)
        ranked_nodes = sorted(
            range(self.nodes_count), key=lambda i: self.ranks[i], reverse=True
        )
        self.transit_nodes = set(ranked_nodes[:num_transit_nodes])

        # 2. Compute Access Nodes for all nodes
        self.forward_access_nodes = [
            self.__compute_access_nodes__(i, True)
            for i in range(self.nodes_count)
        ]
        self.backward_access_nodes = [
            self.__compute_access_nodes__(i, False)
            for i in range(self.nodes_count)
        ]

        # 3. Compute Distance Table between transit nodes
        self.distance_table = {}
        t_nodes_list = list(self.transit_nodes)

        # Note: This is O(T^2) which is slow for Python.
        # For a production implementation, we'd use many-to-one Dijkstra.
        for origin in t_nodes_list:
            # We use CH search for distance
            # This can be optimized by using a simpler Dijkstra if we're only going transit-to-transit
            for target in t_nodes_list:
                if origin == target:
                    self.distance_table[(origin, target)] = 0
                    continue
                # Bidirectional CH search is very fast
                res = CHGraph.search(self, origin, target)
                self.distance_table[(origin, target)] = res["length"]


class TNRGraphAlgorithms:
    def __local_search__(
        self,
        origin_id: int,
        destination_id: int,
        upper_bound: float,
        length_only: bool,
    ) -> Optional[dict]:
        forward_distances = {origin_id: 0}
        forward_parent = {origin_id: -1}
        forward_open_leaves = [(0, origin_id)]

        backward_distances = {destination_id: 0}
        backward_parent = {destination_id: -1}
        backward_open_leaves = [(0, destination_id)]

        best_dist = upper_bound
        meeting_node = -1

        while forward_open_leaves or backward_open_leaves:
            if forward_open_leaves:
                current_distance, current_id = heappop(
                    forward_open_leaves
                )
                if current_distance > best_dist:
                    forward_open_leaves = []
                elif current_id not in self.transit_nodes:
                    current_rank = self.__get_rank__(current_id)
                    neighbors = (
                        self.forward_graph[current_id]
                        if current_id < self.nodes_count
                        else self.original_graph[current_id]
                    )

                    for neighbor_id, weight in neighbors.items():
                        if (
                            self.__get_rank__(neighbor_id) <= current_rank
                            and neighbor_id < self.nodes_count
                        ):
                            continue
                        new_dist = current_distance + weight
                        if new_dist < forward_distances.get(
                            neighbor_id, float("inf")
                        ):
                            forward_distances[neighbor_id] = new_dist
                            if not length_only:
                                forward_parent[neighbor_id] = current_id
                            heappush(
                                forward_open_leaves, (new_dist, neighbor_id)
                            )
                            if (
                                neighbor_id in backward_distances
                                and new_dist + backward_distances[neighbor_id]
                                < best_dist
                            ):
                                best_dist = (
                                    new_dist + backward_distances[neighbor_id]
                                )
                                meeting_node = neighbor_id

            if backward_open_leaves:
                current_distance, current_id = heappop(
                    backward_open_leaves
                )
                if current_distance > best_dist:
                    backward_open_leaves = []
                elif current_id not in self.transit_nodes:
                    current_rank = self.__get_rank__(current_id)
                    neighbors = (
                        self.backward_graph[current_id]
                        if current_id < self.nodes_count
                        else self.original_graph[current_id]
                    )

                    for neighbor_id, weight in neighbors.items():
                        if (
                            self.__get_rank__(neighbor_id) <= current_rank
                            and neighbor_id < self.nodes_count
                        ):
                            continue
                        new_dist = current_distance + weight
                        if new_dist < backward_distances.get(
                            neighbor_id, float("inf")
                        ):
                            backward_distances[neighbor_id] = new_dist
                            if not length_only:
                                backward_parent[neighbor_id] = current_id
                            heappush(
                                backward_open_leaves, (new_dist, neighbor_id)
                            )
                            if (
                                neighbor_id in forward_distances
                                and new_dist + forward_distances[neighbor_id]
                                < best_dist
                            ):
                                best_dist = (
                                    new_dist + forward_distances[neighbor_id]
                                )
                                meeting_node = neighbor_id

            forward_min = (
                forward_open_leaves[0][0]
                if forward_open_leaves
                else float("inf")
            )
            backward_min = (
                backward_open_leaves[0][0]
                if backward_open_leaves
                else float("inf")
            )
            if forward_min > best_dist and backward_min > best_dist:
                break

        if length_only:
            return {"length": best_dist}

        if meeting_node != -1:
            path = self.__reconstruct_ch_path__(
                origin_id,
                destination_id,
                meeting_node,
                forward_parent,
                backward_parent,
            )
            return {"path": path, "length": best_dist}

        return None

    def search(
        self,
        origin_id: int,
        destination_id: int,
        length_only: bool = False,
        **kwargs,
    ) -> dict[str, Any]:
        """
        Function:
        - Perform a TNR search with a highly pruned local CH fallback
        """
        if origin_id == destination_id:
            return (
                {"length": 0}
                if length_only
                else {"path": [origin_id], "length": 0}
            )

        # Access Nodes for origin/destination
        if origin_id < self.nodes_count:
            f_access = self.forward_access_nodes[origin_id]
        else:
            f_access = self.__compute_access_nodes__(origin_id, True)

        if destination_id < self.nodes_count:
            b_access = self.backward_access_nodes[destination_id]
        else:
            b_access = self.__compute_access_nodes__(destination_id, False)

        best_dist = float("inf")

        # Global Query via Distance Table
        for t_f, d_f in f_access.items():
            for t_b, d_b in b_access.items():
                d_table = self.distance_table.get((t_f, t_b), float("inf"))
                total = d_f + d_table + d_b
                if total < best_dist:
                    best_dist = total

        # Locality Filter / Correctness Check
        if length_only:
            return self.__local_search__(
                origin_id, destination_id, best_dist, True
            )

        local_res = self.__local_search__(
            origin_id, destination_id, best_dist, False
        )
        if local_res is not None:
            return local_res

        # Reconstructing path from transit pairs is complex without unpacking tables.
        # Fall back to a standard CH search if global TNR path is needed.
        return CHGraph.search(self, origin_id, destination_id)


class TNRGraph(TNRGraphIO, TNRGraphPreprocessing, TNRGraphAlgorithms, CHGraph):
    def __init__(
        self,
        graph: Optional[list[dict[int, int | float]]] = None,
        num_transit_nodes: int = 100,
        **kwargs,
    ):
        """
        Function:
        - Create a Transit Node Routing Graph object

        Optional Arguments:
        - `graph`: Original graph adjacency list
        - `num_transit_nodes`: Number of top-ranked nodes to use as transit nodes
        - `kwargs`: Used when loading from saved state
        """
        if "transit_nodes" in kwargs:
            # Loading from saved state
            self.nodes_count = kwargs.get("nodes_count")
            self.transit_nodes = set(kwargs.get("transit_nodes"))

            # Reconstruct distance_table (keys are strings like "(1,2)")
            self.distance_table = {}
            for k, v in kwargs.get("distance_table").items():
                pair = tuple(map(int, k.strip("()").split(",")))
                self.distance_table[pair] = v

            # Reconstruct access nodes
            self.forward_access_nodes = [
                {int(k): v for k, v in d.items()}
                for d in kwargs.get("forward_access_nodes")
            ]
            self.backward_access_nodes = [
                {int(k): v for k, v in d.items()}
                for d in kwargs.get("backward_access_nodes")
            ]

            # Initialize CH part
            ch_data = kwargs.get("ch_data")
            # Convert graph keys to integers
            for key in ["forward_graph", "backward_graph", "original_graph"]:
                if ch_data.get(key) is not None:
                    ch_data[key] = [
                        {int(k): v for k, v in d.items()} for d in ch_data[key]
                    ]
            super().__init__(**ch_data)
        else:
            # Preprocess from original graph
            super().__init__(graph=graph)
            self.__preprocess_tnr__(num_transit_nodes)

    def get_shortest_path(
        self, origin_id: int, destination_id: int, **kwargs: Any
    ) -> dict[str, Any]:
        return self.search(origin_id, destination_id)

File: scgraph/cpp/src/transit_node_routing.cpp
#include "transit_node_routing.hpp"
#include <queue>
#include <algorithm>
#include <limits>
#include <iostream>
#include <functional>
#include <thread>

TNRGraph::TNRGraph(const std::vector<std::unordered_map<int, double>>& graph,
                   int settled_limit,
                   int num_transit_nodes,
                   std::function<double(CHGraph*, int)> heuristic_fn)
    : CHGraph(graph, settled_limit, heuristic_fn) {
    // 1. Select Transit Nodes
    std::vector<int> sorted_nodes(nodes_count);
    for (int i = 0; i < nodes_count; ++i) {
        sorted_nodes[i] = i;
    }
    std::sort(sorted_nodes.begin(), sorted_nodes.end(), [this](int a, int b) {
        return ranks[a] > ranks[b];
    });

    int actual_num_transit = std::min(num_transit_nodes, nodes_count);
    for (int i = 0; i < actual_num_transit; ++i) {
        transit_nodes.insert(sorted_nodes[i]);
    }

    // 2. Compute Access Nodes in Parallel
    forward_access_nodes.assign(nodes_count, {});
    backward_access_nodes.assign(nodes_count, {});

    std::vector<bool> is_transit(nodes_count, false);
    for (int tn : transit_nodes) {
        is_transit[tn] = true;
    }

    unsigned int num_threads = std::thread::hardware_concurrency();
    if (num_threads == 0) num_threads = 4;

    std::vector<std::thread> threads;
    threads.reserve(num_threads);

    for (unsigned int thread_idx = 0; thread_idx < num_threads; ++thread_idx) {
        threads.emplace_back([this, thread_idx, num_threads, &is_transit]() {
            std::vector<double> distances(nodes_count, std::numeric_limits<double>::infinity());
            std::vector<int> visited_nodes;

            auto local_compute_access_nodes = [this, &distances, &visited_nodes, &is_transit](int node_id, bool forward) {
                std::unordered_map<int, double> access_nodes;
                distances[node_id] = 0.0;
                visited_nodes.push_back(node_id);
                using PQItem = std::pair<double, int>;
                std::priority_queue<PQItem, std::vector<PQItem>, std::greater<PQItem>> open_leaves;
                open_leaves.push({0.0, node_id});

                while (!open_leaves.empty()) {
                    auto [dist, current_id] = open_leaves.top();
                    open_leaves.pop();

                    if (dist > distances[current_id]) {
                        continue;
                    }

                    if (is_transit[current_id]) {
                        if (access_nodes.find(current_id) == access_nodes.end() || dist < access_nodes[current_id]) {
                            access_nodes[current_id] = dist;
                        }
                        continue;
                    }

                    const auto& neighbors = forward ? forward_graph[current_id] : backward_graph[current_id];
                    for (const auto& [neighbor_id, weight] : neighbors) {
                        double new_dist = dist + weight;
                        if (new_dist < distances[neighbor_id]) {
                            if (distances[neighbor_id] == std::numeric_limits<double>::infinity()) {
                                visited_nodes.push_back(neighbor_id);
                            }
                            distances[neighbor_id] = new_dist;
                            open_leaves.push({new_dist, neighbor_id});
                        }
                    }
                }

                for (int v : visited_nodes) {
                    distances[v] = std::numeric_limits<double>::infinity();
                }
                visited_nodes.clear();

                return access_nodes;
            };

            for (int i = thread_idx; i < nodes_count; i += num_threads) {
                forward_access_nodes[i] = local_compute_access_nodes(i, true);
                backward_access_nodes[i] = local_compute_access_nodes(i, false);
            }
        });
    }

    for (auto& t : threads) {
        if (t.joinable()) {
            t.join();
        }
    }

    // 3. Compute Distance Table in Parallel
    num_transit = transit_nodes.size();
    transit_node_to_local_idx.assign(nodes_count, -1);
    std::vector<int> transit_nodes_vec(transit_nodes.begin(), transit_nodes.end());
    for (int i = 0; i < num_transit; ++i) {
        transit_node_to_local_idx[transit_nodes_vec[i]] = i;
    }
    distance_table_flat.assign(num_transit * num_transit, std::numeric_limits<double>::infinity());

    std::vector<std::thread> dt_threads;
    dt_threads.reserve(num_threads);

    for (unsigned int thread_idx = 0; thread_idx < num_threads; ++thread_idx) {
        dt_threads.emplace_back([this, thread_idx, num_threads, &transit_nodes_vec]() {
            size_t n = original_graph.size();
            std::vector<double> dist(n, std::numeric_limits<double>::infinity());
            std::vector<int> visited;

            for (int origin_idx = thread_idx; origin_idx < num_transit; origin_idx += num_threads) {
                int origin = transit_nodes_vec[origin_idx];
                dist[origin] = 0.0;
                visited.push_back(origin);
                using PQItem = std::pair<double, int>;
                std::priority_queue<PQItem, std::vector<PQItem>, std::greater<PQItem>> pq;
                pq.push({0.0, origin});
                while (!pq.empty()) {
                    auto [d, u] = pq.top();
                    pq.pop();
                    if (d > dist[u]) continue;
                    for (const auto& [v, w] : original_graph[u]) {
                        double nd = d + w;
                        if (nd < dist[v]) {
                            if (dist[v] == std::numeric_limits<double>::infinity()) {
                                visited.push_back(v);
                            }
                            dist[v] = nd;
                            pq.push({nd, v});
                        }
                    }
                }
                for (int target_idx = 0; target_idx < num_transit; ++target_idx) {
                    int target = transit_nodes_vec[target_idx];
                    distance_table_flat[origin_idx * num_transit + target_idx] = dist[target];
                }
                for (int v : visited) {
                    dist[v] = std::numeric_limits<double>::infinity();
                }
                visited.clear();
            }
        });
    }

    for (auto& t : dt_threads) {
        if (t.joinable()) {
            t.join();
        }
    }

    for (int i = 0; i < num_transit; ++i) {
        for (int j = 0; j < num_transit; ++j) {
            distance_table[{transit_nodes_vec[i], transit_nodes_vec[j]}] = distance_table_flat[i * num_transit + j];
        }
    }
}

TNRGraph::TNRGraph(int nodes_count,
                   const std::vector<int>& ranks,
                   const std::vector<std::unordered_map<int, double>>& forward_graph,
                   const std::vector<std::unordered_map<int, double>>& backward_graph,
                   const std::unordered_map<std::pair<int, int>, int, pair_hash>& shortcuts,
                   const std::optional<std::vector<std::unordered_map<int, double>>>& original_graph,
                   const std::set<int>& transit_nodes,
                   const std::unordered_map<std::pair<int, int>, double, pair_hash>& distance_table,
                   const std::vector<std::unordered_map<int, double>>& forward_access_nodes,
                   const std::vector<std::unordered_map<int, double>>& backward_access_nodes,
                   int settled_limit)
    : CHGraph(nodes_count, ranks, forward_graph, backward_graph, shortcuts, original_graph, settled_limit),
      transit_nodes(transit_nodes), distance_table(distance_table),
      forward_access_nodes(forward_access_nodes), backward_access_nodes(backward_access_nodes) {
    initialize_fast_lookup();
}

void TNRGraph::initialize_fast_lookup() {
    num_transit = transit_nodes.size();
    transit_node_to_local_idx.assign(nodes_count, -1);
    int idx = 0;
    for (int node : transit_nodes) {
        if (node >= 0 && node < nodes_count) {
            transit_node_to_local_idx[node] = idx++;
        }
    }
    distance_table_flat.assign(num_transit * num_transit, std::numeric_limits<double>::infinity());
    for (const auto& [pair, dist] : distance_table) {
        int u = (pair.first >= 0 && pair.first < nodes_count) ? transit_node_to_local_idx[pair.first] : -1;
        int v = (pair.second >= 0 && pair.second < nodes_count) ? transit_node_to_local_idx[pair.second] : -1;
        if (u != -1 && v != -1) {
            distance_table_flat[u * num_transit + v] = dist;
        }
    }
}

std::optional<GraphResult> TNRGraph::local_search(int origin_id, int destination_id, double upper_bound, bool length_only) const {
    int max_node_id = std::max(origin_id, destination_id);
    int current_sz = static_cast<int>(query_f_distances.size());
    if (max_node_id >= current_sz) {
        int new_sz = max_node_id + 1;
        query_f_distances.resize(new_sz, std::numeric_limits<double>::infinity());
        query_b_distances.resize(new_sz, std::numeric_limits<double>::infinity());
        query_f_parents.resize(new_sz, -1);
        query_b_parents.resize(new_sz, -1);
    }

    query_f_distances[origin_id] = 0.0;
    if (!length_only) query_f_parents[origin_id] = -1;
    query_visited.push_back(origin_id);

    query_b_distances[destination_id] = 0.0;
    if (!length_only) query_b_parents[destination_id] = -1;
    query_visited.push_back(destination_id);

    using PQItem = std::pair<double, int>;
    std::priority_queue<PQItem, std::vector<PQItem>, std::greater<PQItem>> forward_open_leaves, backward_open_leaves;
    forward_open_leaves.push({0.0, origin_id});
    backward_open_leaves.push({0.0, destination_id});

    double best_dist = upper_bound;
    int meeting_node = -1;

    while (!forward_open_leaves.empty() || !backward_open_leaves.empty()) {
        if (!forward_open_leaves.empty()) {
            auto [current_distance, current_id] = forward_open_leaves.top();
            forward_open_leaves.pop();

            if (current_distance > best_dist) {
                while (!forward_open_leaves.empty()) forward_open_leaves.pop();
            } else if (current_id < 0 || current_id >= nodes_count || transit_node_to_local_idx[current_id] == -1) {
                if (current_id < nodes_count) {
                    for (const auto& [neighbor_id, weight] : forward_graph[current_id]) {
                        double new_dist = current_distance + weight;
                        if (neighbor_id >= static_cast<int>(query_f_distances.size())) {
                            int new_sz = neighbor_id + 1;
                            query_f_distances.resize(new_sz, std::numeric_limits<double>::infinity());
                            query_b_distances.resize(new_sz, std::numeric_limits<double>::infinity());
                            query_f_parents.resize(new_sz, -1);
                            query_b_parents.resize(new_sz, -1);
                        }
                        if (query_f_distances[neighbor_id] == std::numeric_limits<double>::infinity()) {
                            query_visited.push_back(neighbor_id);
                        }
                        if (new_dist < query_f_distances[neighbor_id]) {
                            query_f_distances[neighbor_id] = new_dist;
                            if (!length_only) query_f_parents[neighbor_id] = current_id;
                            forward_open_leaves.push({new_dist, neighbor_id});
                            if (query_b_distances[neighbor_id] != std::numeric_limits<double>::infinity() && new_dist + query_b_distances[neighbor_id] < best_dist) {
                                best_dist = new_dist + query_b_distances[neighbor_id];
                                meeting_node = neighbor_id;
                            }
                        }
                    }
                } else {
                    double current_rank = get_rank(current_id);
                    for (const auto& [neighbor_id, weight] : original_graph[current_id]) {
                        double neighbor_rank = get_rank(neighbor_id);
                        if (neighbor_rank <= current_rank && neighbor_id < nodes_count) continue;

                        double new_dist = current_distance + weight;
                        if (neighbor_id >= static_cast<int>(query_f_distances.size())) {
                            int new_sz = neighbor_id + 1;
                            query_f_distances.resize(new_sz, std::numeric_limits<double>::infinity());
                            query_b_distances.resize(new_sz, std::numeric_limits<double>::infinity());
                            query_f_parents.resize(new_sz, -1);
                            query_b_parents.resize(new_sz, -1);
                        }
                        if (query_f_distances[neighbor_id] == std::numeric_limits<double>::infinity()) {
                            query_visited.push_back(neighbor_id);
                        }
                        if (new_dist < query_f_distances[neighbor_id]) {
                            query_f_distances[neighbor_id] = new_dist;
                            if (!length_only) query_f_parents[neighbor_id] = current_id;
                            forward_open_leaves.push({new_dist, neighbor_id});
                            if (query_b_distances[neighbor_id] != std::numeric_limits<double>::infinity() && new_dist + query_b_distances[neighbor_id] < best_dist) {
                                best_dist = new_dist + query_b_distances[neighbor_id];
                                meeting_node = neighbor_id;
                            }
                        }
                    }
                }
            }
        }

        if (!backward_open_leaves.empty()) {
            auto [current_distance, current_id] = backward_open_leaves.top();
            backward_open_leaves.pop();

            if (current_distance > best_dist) {
                while (!backward_open_leaves.empty()) backward_open_leaves.pop();
            } else if (current_id < 0 || current_id >= nodes_count || transit_node_to_local_idx[current_id] == -1) {
                if (current_id < nodes_count) {
                    for (const auto& [neighbor_id, weight] : backward_graph[current_id]) {
                        double new_dist = current_distance + weight;
                        if (neighbor_id >= static_cast<int>(query_f_distances.size())) {
                            int new_sz = neighbor_id + 1;
                            query_f_distances.resize(new_sz, std::numeric_limits<double>::infinity());
                            query_b_distances.resize(new_sz, std::numeric_limits<double>::infinity());
                            query_f_parents.resize(new_sz, -1);
                            query_b_parents.resize(new_sz, -1);
                        }
                        if (query_b_distances[neighbor_id] == std::numeric_limits<double>::infinity()) {
                            query_visited.push_back(neighbor_id);
                        }
                        if (new_dist < query_b_distances[neighbor_id]) {
                            query_b_distances[neighbor_id] = new_dist;
                            if (!length_only) query_b_parents[neighbor_id] = current_id;
                            backward_open_leaves.push({new_dist, neighbor_id});
                            if (query_f_distances[neighbor_id] != std::numeric_limits<double>::infinity() && new_dist + query_f_distances[neighbor_id] < best_dist) {
                                best_dist = new_dist + query_f_distances[neighbor_id];
                                meeting_node = neighbor_id;
                            }
                        }
                    }
                } else {
                    double current_rank = get_rank(current_id);
                    for (const auto& [neighbor_id, weight] : original_graph[current_id]) {
                        double neighbor_rank = get_rank(neighbor_id);
                        if (neighbor_rank <= current_rank && neighbor_id < nodes_count) continue;

                        double new_dist = current_distance + weight;
                        if (neighbor_id >= static_cast<int>(query_f_distances.size())) {
                            int new_sz = neighbor_id + 1;
                            query_f_distances.resize(new_sz, std::numeric_limits<double>::infinity());
                            query_b_distances.resize(new_sz, std::numeric_limits<double>::infinity());
                            query_f_parents.resize(new_sz, -1);
                            query_b_parents.resize(new_sz, -1);
                        }
                        if (query_b_distances[neighbor_id] == std::numeric_limits<double>::infinity()) {
                            query_visited.push_back(neighbor_id);
                        }
                        if (new_dist < query_b_distances[neighbor_id]) {
                            query_b_distances[neighbor_id] = new_dist;
                            if (!length_only) query_b_parents[neighbor_id] = current_id;
                            backward_open_leaves.push({new_dist, neighbor_id});
                            if (query_f_distances[neighbor_id] != std::numeric_limits<double>::infinity() && new_dist + query_f_distances[neighbor_id] < best_dist) {
                                best_dist = new_dist + query_f_distances[neighbor_id];
                                meeting_node = neighbor_id;
                            }
                        }
                    }
                }
            }
        }

        double forward_min = forward_open_leaves.empty() ? std::numeric_limits<double>::infinity() : forward_open_leaves.top().first;
        double backward_min = backward_open_leaves.empty() ? std::numeric_limits<double>::infinity() : backward_open_leaves.top().first;
        if (forward_min > best_dist && backward_min > best_dist) break;
    }

    std::optional<GraphResult> res = std::nullopt;
    if (meeting_node != -1) {
        if (length_only) {
            res = GraphResult{{}, best_dist};
        } else {
            std::vector<int> path = reconstruct_ch_path(origin_id, destination_id, meeting_node, query_f_parents, query_b_parents);
            res = GraphResult{path, best_dist};
        }
    }

    for (int v : query_visited) {
        query_f_distances[v] = std::numeric_limits<double>::infinity();
        query_b_distances[v] = std::numeric_limits<double>::infinity();
        query_f_parents[v] = -1;
        query_b_parents[v] = -1;
    }
    query_visited.clear();

    return res;
}

GraphResult TNRGraph::search(int origin_id, int destination_id, bool length_only) const {
    if (origin_id == destination_id) {
        return {length_only ? std::vector<int>{} : std::vector<int>{origin_id}, 0.0};
    }

    std::unordered_map<int, double> f_access_temp, b_access_temp;
    const std::unordered_map<int, double>* f_access = nullptr;
    const std::unordered_map<int, double>* b_access = nullptr;

    // Forward Access Nodes
    if (origin_id < nodes_count) {
        f_access = &forward_access_nodes[origin_id];
    } else {
        // Compute for added node
        std::unordered_map<int, double> distances;
        distances[origin_id] = 0.0;
        using PQItem = std::pair<double, int>;
        std::priority_queue<PQItem, std::vector<PQItem>, std::greater<PQItem>> open_leaves;
        open_leaves.push({0.0, origin_id});
        while (!open_leaves.empty()) {
            auto [current_distance, current_id] = open_leaves.top();
            open_leaves.pop();
            if (current_id >= 0 && current_id < nodes_count && transit_node_to_local_idx[current_id] != -1) {
                if (f_access_temp.find(current_id) == f_access_temp.end() || current_distance < f_access_temp[current_id]) {
                    f_access_temp[current_id] = current_distance;
                }
                continue;
            }
            double current_rank = get_rank(current_id);
            const auto& neighbors = (current_id < nodes_count) ? forward_graph[current_id] : original_graph[current_id];
            for (const auto& [neighbor_id, weight] : neighbors) {
                if (current_id < nodes_count && get_rank(neighbor_id) <= current_rank && neighbor_id < nodes_count) continue;
                double new_dist = current_distance + weight;
                if (distances.find(neighbor_id) == distances.end() || new_dist < distances[neighbor_id]) {
                    distances[neighbor_id] = new_dist;
                    open_leaves.push({new_dist, neighbor_id});
                }
            }
        }
        f_access = &f_access_temp;
    }

    // Backward Access Nodes
    if (destination_id < nodes_count) {
        b_access = &backward_access_nodes[destination_id];
    } else {
        // Compute for added node
        std::unordered_map<int, double> distances;
        distances[destination_id] = 0.0;
        using PQItem = std::pair<double, int>;
        std::priority_queue<PQItem, std::vector<PQItem>, std::greater<PQItem>> open_leaves;
        open_leaves.push({0.0, destination_id});
        while (!open_leaves.empty()) {
            auto [current_distance, current_id] = open_leaves.top();
            open_leaves.pop();
            if (current_id >= 0 && current_id < nodes_count && transit_node_to_local_idx[current_id] != -1) {
                if (b_access_temp.find(current_id) == b_access_temp.end() || current_distance < b_access_temp[current_id]) {
                    b_access_temp[current_id] = current_distance;
                }
                continue;
            }
            double current_rank = get_rank(current_id);
            const auto& neighbors = (current_id < nodes_count) ? backward_graph[current_id] : original_graph[current_id];
            for (const auto& [neighbor_id, weight] : neighbors) {
                if (current_id < nodes_count && get_rank(neighbor_id) <= current_rank && neighbor_id < nodes_count) continue;
                double new_dist = current_distance + weight;
                if (distances.find(neighbor_id) == distances.end() || new_dist < distances[neighbor_id]) {
                    distances[neighbor_id] = new_dist;
                    open_leaves.push({new_dist, neighbor_id});
                }
            }
        }
        b_access = &b_access_temp;
    }

    double best_dist = std::numeric_limits<double>::infinity();
    for (const auto& [t_f, d_f] : *f_access) {
        int u = (t_f >= 0 && t_f < nodes_count) ? transit_node_to_local_idx[t_f] : -1;
        if (u == -1) continue;
        for (const auto& [t_b, d_b] : *b_access) {
            int v = (t_b >= 0 && t_b < nodes_count) ? transit_node_to_local_idx[t_b] : -1;
            if (v == -1) continue;
            double d_table = distance_table_flat[u * num_transit + v];
            if (d_table != std::numeric_limits<double>::infinity()) {
                double total = d_f + d_table + d_b;
                if (total < best_dist) {
                    best_dist = total;
                }
            }
        }
    }

    auto local_res = local_search(origin_id, destination_id, best_dist, length_only);
    if (local_res.has_value()) {
        return local_res.value();
    }

    if (length_only) {
        return {{}, best_dist};
    }

    // Fallback to CH search for path reconstruction if global TNR path was found but local search failed
    return CHGraph::search(origin_id, destination_id);
}
